### ISSUE TITLE
pnext: Unstable process management back-end

### ISSUE THREAD

**Issue by eriohl** [Starcounter/Starcounter#3199]
**scdata.exe** and **scdblog.exe** can too easily get out of sync foiling startup. Needs to be looked over.

**Comment by alemoi** [comment]
I have pnext and tried to start an empty database but can't because of the following:
```
Starcounter.DbException: ScErrDbProcTerminated (SCERR10028): The server detected that the database data process unexpectedly terminated during a management operation. scdata.exe, PID=13084, Database=default, Exitcode=4256
Version: 2.0.0.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR10028.
   at Starcounter.Server.DatabaseEngine.StartDatabaseProcess(Database database) in c:\github\Starcounter\Level1\src\Server\Starcounter.Server\CoreServices\DatabaseEngine.cs:line 170
   at Starcounter.Server.Commands.Processors.StartDatabaseCommandProcessor.<>c__DisplayClass2_0.<Execute>b__0(CommandTask task) in c:\github\Starcounter\Level1\src\Server\Starcounter.Server\Commands\Processors\StartDatabaseCommandProcessor.cs:line 54
   at Starcounter.Server.Commands.CommandProcessor.WithinTask(CommandTask task, Func`2 func) in c:\github\Starcounter\Level1\src\Server\Starcounter.Server\Commands\CommandProcessor.cs:line 510
   at Starcounter.Server.Commands.Processors.StartDatabaseCommandProcessor.Execute() in c:\github\Starcounter\Level1\src\Server\Starcounter.Server\Commands\Processors\StartDatabaseCommandProcessor.cs:line 43
   at Starcounter.Server.Commands.CommandProcessor.ProcessCommand(NotifyCommandStatusChangedCallback notifyStatusChangedCallback) in c:\github\Starcounter\Level1\src\Server\Starcounter.Server\Commands\CommandProcessor.cs:line 298
ErrorCode=10028
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR10028
---> Starcounter.DbException: ScErrTransLogWriterNotAvailable (SCERR4256): Transaction log writer is not available.
Version: 2.0.0.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR4256.
ErrorCode=4256
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR4256
```

Is it related to this issue or something else? Thank you.

**Comment by eriohl** [comment]
Possibly. Depends on the underlying reason for the failure, which isn't very clear from this.

**Comment by alemoi** [comment]
Unfortunately there is nothing else in the log: A fatal error was detected: ScErrTransLogWriterNotAvailable (SCERR4256): Transaction log writer is not available.

**Comment by eriohl** [comment]
I'm thinking that we should have scdblog.exe start scdata.exe instead of the other way around. Then establishing a connection from scdata.exe to scdblog.exe will be a lot more stable because we don't need to handle it with a timeout. Additionally: #3195 will be easy to address as well using job object (as for logopt). I think it will be more correct also since scdblog.exe is to outlast scdata.exe. What do you think @bigwad?

**Comment by bigwad** [comment]
It'll simplify process management a lot, but not so consistent on architecture design. Say, it makes harder to launch readonly db instance that doesn't require scdblog.

**Comment by eriohl** [comment]
If you want to run without the log writer you can just launch scdata.exe directly. Or scdblog.exe can just not start the log writer thread (running scdblog.exe doesn't actually start the log writer, it is scdate.exe connecting to it that does). So I I don't think running without log writing will be an issue.

As for whether it is consistent with the architecture design I it can of hard to say since this work was suspended. I don't see why scdata.exe starting scdblog.exe fits much better?

**Comment by bigwad** [comment]
On running without logwriter - I mean that it would be perfect if switching to readonly mode would be as simple as passing "-readonly" parameter to db executable. So nobody would care running another executable.

On architecture -  I believe it's more consistent if scdata launches scdblog (as now) as scdata requires scdblog's services, not vice versa. Or some upper level component should exist that launches and ties together both of them depending on db mode. The only question - how to implement it without bringing unnecessary complexity.

**Comment by eriohl** [comment]
Since the actual log writer is launched only when the scdata.exe connects, having scdblog.exe launching scdata.exe is pretty much the same as having a third process launching them both (with the difference that the log writer runs in a separate thread rather then a separate process). I agree that it feels better that scdata.exe launches scdblog.exe (which is why I did this to begin with) but I think this issue can be solved simply by looking at it differently (maybe rename scdblog.exe also).

**Comment by bigwad** [comment]
I agree, that we can look at it as scdblog is the main db process. Then we shouldn't run scdata on its own, but only via scdblog (renamed).
Though the better solution from my point of view would be to launch scdata and scdblog piped (scdata|scdblog) under the supervision of some upper process. Maybe it has some problems from performance point of view - has to be probed.
But the simplest solution - is just as you proposed. No doubt.

**Comment by dan31** [comment]
Thanks to @chrhol and @andwah we now produce PNextNightly build with installer downloadable http://downloads.starcounter.com/download. Having installed Starcounter with defaults, I get the same behaviour when run any Starcounter application from Visual Studio on a virtual machine:
> The server detected that the database data process unexpectedly terminated during a management operation. scdata.exe, PID=7096, Database=default, Exitcode=4117 (ScErrDbProcTerminated (SCERR10028))

What's the difference from you guys running to Starcounter built locally from source code to Starcounter installed from the installer, maybe there is just something minimal missing. @alemoi, the tests that require having Starcounter installed, do any of them pass? If they do, that is quite strange, but this might be an evidence for the current issue.

**Comment by alemoi** [comment]
Most of the tests are running in the environment that is equal to installation, except for whitelists, for example.

**Comment by chrhol** [comment]
> What's the difference from you guys running to Starcounter built locally from source code to Starcounter installed from the installer

We found the problem, not related to this in this case. @dan31 had an old database in the repo so a new one was not created, and the imagefiles could not be found.

**Comment by dan31** [comment]
My false alarm, I simply forgot I haven't removed the old database dir. It worked well now, whilst the error message today is too generic. We will have a separate issue on its improvements. @alemoi, can you now try to install and run pnext from the download website as we done last time to try replicator?

**Comment by chrhol** [comment]
Some builds on the buildserver fails (in different tests) with error `SCERR4256: Transaction log writer is not available`. Are these failures something that are connected with this issue? 

These are from two different builds (2.3.521 and 2.3.505), and the error occurs in two different tests that should succeed:

```
[14:21:37][Step 1/1] CommitHooks_test-bat (1m:52s)
[14:21:37][CommitHooks_test-bat] -- Running process "C:\TeamCity\TeamCity9\buildAgent\work\sc-1506\Level1\src\Samples\TestApps\CommitHooksTest\CommitHooks_test.bat "
[14:21:37][CommitHooks_test-bat] Running CommitHooksTest regression test.
[14:21:40][CommitHooks_test-bat] 
[14:21:41][CommitHooks_test-bat] [14:21:40, Critical, Starcounter (commithookstest)]
[14:21:41][CommitHooks_test-bat] A fatal error was detected: ScErrTransLogWriterNotAvailable (SCERR4256): Transaction log writer is not available.
[14:21:41][CommitHooks_test-bat] 
[14:21:41][CommitHooks_test-bat] [14:21:40, Error, Starcounter (commithookstest)]
[14:21:41][CommitHooks_test-bat] ScErrUnexpPlatformAPIError (SCERR6067): An unexpected error occured calling a platform API function. OS error=535 (ConnectNamedPipe).
[14:21:41][CommitHooks_test-bat] 
[14:21:41][CommitHooks_test-bat] Error: CommitHooksTest failed!
[14:21:41][CommitHooks_test-bat] Test failed: "CommitHooks_test.bat" at "C:\TeamCity\TeamCity9\buildAgent\work\sc-1506\Level1\src\Samples\TestApps\CommitHooksTest\CommitHooks_test.bat" with error code 1.
```

```
[11:18:11][Step 1/1] OffsetkeyBug2915_test-bat (1m:11s)
[11:18:11][OffsetkeyBug2915_test-bat] -- Running process "C:\TeamCity\TeamCity9\buildAgent\work\sc-1490\Level1\src\Samples\TestApps\OffsetkeyBug2915\OffsetkeyBug2915_test.bat "
[11:18:11][OffsetkeyBug2915_test-bat] Running OffsetkeyBug2915 regression test.
[11:18:11][OffsetkeyBug2915_test-bat] Run the test
[11:18:15][OffsetkeyBug2915_test-bat] 
[11:18:15][OffsetkeyBug2915_test-bat] [11:18:15, Critical, Starcounter (offsetkeybug2915db)]
[11:18:15][OffsetkeyBug2915_test-bat] A fatal error was detected: ScErrTransLogWriterNotAvailable (SCERR4256): Transaction log writer is not available.
[11:18:15][OffsetkeyBug2915_test-bat] 
[11:18:15][OffsetkeyBug2915_test-bat] [11:18:15, Error, Starcounter (offsetkeybug2915db)]
[11:18:15][OffsetkeyBug2915_test-bat] ScErrUnexpPlatformAPIError (SCERR6067): An unexpected error occured calling a platform API function. OS error=535 (ConnectNamedPipe).
[11:18:15][OffsetkeyBug2915_test-bat] 
[11:18:15][OffsetkeyBug2915_test-bat] Error: OffsetkeyBug2915 failed!
[11:18:15][OffsetkeyBug2915_test-bat] Test failed: "OffsetkeyBug2915_test.bat" at "C:\TeamCity\TeamCity9\buildAgent\work\sc-1490\Level1\src\Samples\TestApps\OffsetkeyBug2915\OffsetkeyBug2915_test.bat" with error code 1.
```

**Comment by dan31** [comment]
@eriohl @bigwad What is the status for this one? I discussed with @chrhol and @k-rus, this appears to be most impeding and is still recurring. It also seems to me that we missed that one throughout email discussion.

**Comment by bigwad** [comment]
@dan31, I've just mentioned it in the planning e-mail thread

**Comment by eriohl** [comment]
I didn't see that people was that much impeded by build server tests failing on occasion?

**Comment by k-rus** [comment]
> I didn't see that people was that much impeded by build server tests failing on occasion?

I met 4 failures of the builds due to the issue just within few hours. In my opinion it happens too often.

**Comment by eriohl** [comment]
Ok. @bigwad: If you could look at this it would be great?

**Comment by bigwad** [comment]
Have fixed the most often reason for the error, so I hope it won't be a problem on the builder anymore. Will think about redesign we discussed with @eriohl to make it rock solid in all cases.

**Comment by eriohl** [comment]
Mentioned earlier. This:

```
sccreatedb.exe -ip c:/temp/testapp1 testapp1
sccreatedb.exe -ip c:/temp/testapp1 testapp2

start scdata.exe 1 TESTAPP1 c:/temp/testapp1 TESTAPP1 c:/temp/testapp1 c:/temp/testapp1
start scdata.exe 1 TESTAPP2 c:/temp/testapp1 TESTAPP2 c:/temp/testapp1 c:/temp/testapp1
```

Causes an assertion failure in one of the scdata.exe instances (rather then a nice error) because the instance id for both instances are the same.

**Comment by eriohl** [comment]
If not an immediate issue anymore we should discuss this in relation to overall architecture and do this properly.

As it is stands now we will be running one DM per database (although perhaps not one log writer) rather then the single one previous discussed for the foreseeable future (i e pretty much as in previous versions). So we need a monitor process as well regardless. (We do have one in current version generally referred to as the administrator but we need an additional one regardless because they must be able to run in different sessions.)

So I think we should to the third process and have the administrator interact with this one (via pipe or something) to start and stop processes.

But let's have a discussion on this before we proceed with patching this stuff any further so we can move this in the right direction.

**Comment by eriohl** [comment]
I changed so that all shared object names, including control event, are derived from instance id instead of sometimes instance id and sometimes database name. Furthermore I fixed so that a file lock is set on CFG file loading the database. This prevents multiple instances with the same instance id from being booted up and for multiple instance to access the same database (it does however allow multiple databases to exist in different locations).
